bank/api/middleware.go

```diff
@@ -0,0 +1,167 @@
+package api
+
+import (
+	"encoding/json"
+	"github.com/gorilla/context"
+	"github.com/iliyanmotovski/bankv2/bank/domain"
+	"log"
+	"net/http"
+	"time"
+)
+
+func CheckValidSessionOnLoginAndLogoutMiddleware(sessStore domain.SessionStore) func(http.Handler) http.Handler {
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var u domain.User
+			const loginUrl = "/users/login"
+			const logoutUrl = "/users/logout"
+
+			json.NewDecoder(r.Body).Decode(&u)
+			cookie, err := r.Cookie("session")
+			if err != nil {
+				cookie = &http.Cookie{Value: "rfBd56ti2SMtY"}
+			}
+
+			session, ok := sessStore.FindSessionAvailableAt(cookie.Value, time.Now().Local())
+			if ok {
+				switch r.URL.String() {
+				case loginUrl:
+					http.Error(w, "session not expired", 400)
+					return
+				case logoutUrl:
+					context.Set(r, "session", session)
+					next.ServeHTTP(w, r)
+					return
+				}
+			} else if r.URL.String() == logoutUrl {
+				http.Error(w, "already logged out", 400)
+				return
+			}
+
+			context.Set(r, "user", u)
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func CheckValidSessionOnTransactionMiddleware(sessStore domain.SessionStore, accStore domain.AccountStore, renew domain.Renew) func(http.Handler) http.Handler {
```

Comment:
I'm not sure that you need this middleware at all. Single middleware need to be used for security.